### PR TITLE
Telink B91 Flash Driver Improvements 

### PR DESCRIPTION
**Pull request by yurvyn**

### CODE REVIEW

File: drivers/flash/soc_flash_b91.c [hunk 1]
```diff
@@ -104,6 +135,21 @@ static int flash_b91_write(const struct device *dev, off_t offset,
 		return -EINVAL;
 	}
 
+	/* need to store data in intermediate RAM buffer in case from flash to flash write */
+	if ((uint32_t)data > FLASH_ORIGIN) {
+
+		buf = k_malloc(len);
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_b91.c:141]: I would add a limit for memory allocation size and do write of data bigger than data in a few chunks.
This will reduce possibility of encountering memory unavailable issues while writing for flash to flash.

**Review by nvlsianpu — COMMENTED** [review]
The erase improvement looks fine.
I have only suggestion to the F2F write algo.

File: drivers/flash/soc_flash_b91.c (hunk 1 above)
Comment (yurvyn) [reply drivers/flash/soc_flash_b91.c:141]: It makes sense.  However, I prefer to leave it as it is because even if check for CONFIG_HEAP_MEM_POOL_SIZE the k_malloc function still can return NULL, and users still need to do this chunking on the upper layer (or increase CONFIG_HEAP_MEM_POOL_SIZE).
Are you OK with it, or is it something that must be fixed?

Comment (yurvyn) [reply drivers/flash/soc_flash_b91.c:141]: Yes, with the local variable approach the non-mem issue is resolved.
However, I still prefer to avoid chunking in the driver to avoid potential performance issues.

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig [hunk 2]
```diff
@@ -4,9 +4,11 @@
 CONFIG_SOC_SERIES_RISCV_TELINK_B91=y
```
Comment (edersondisouza) [inline boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig:4]: Maybe add in the commit message why this change?

File: drivers/flash/soc_flash_b91.c [hunk 3]
```diff
@@ -76,10 +87,29 @@ static int flash_b91_erase(const struct device *dev, off_t offset, size_t len)
 		return -EACCES;
 	}
 
-	/* erase flash page by page */
-	for (int i = 0; i < page_nums; i++) {
-		flash_erase_page(offset);
-		offset += PAGE_SIZE;
+	while (page_nums) {
+		/* check for 64K erase possibilite, then check for 32K and so on.. */
```
Comment (edersondisouza) [inline drivers/flash/soc_flash_b91.c:91]: s/possibilite/possibility/

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig (hunk 2 above)
Comment (yurvyn) [reply boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig:4]: Done.

File: drivers/flash/soc_flash_b91.c (hunk 3 above)
Comment (yurvyn) [reply drivers/flash/soc_flash_b91.c]: Done.

File: drivers/flash/soc_flash_b91.c (hunk 1 above)
Comment (nvlsianpu) [reply drivers/flash/soc_flash_b91.c:141]: Ok, you know the limitation.

**Review by nvlsianpu — DISMISSED** [review]

File: drivers/flash/soc_flash_b91.c [hunk 4]
```diff
@@ -104,6 +135,21 @@ static int flash_b91_write(const struct device *dev, off_t offset,
...
+		/* copy Flash data to RAM */
+		memcpy(buf, data, len);
+
+		/* substitute data with allocated buffer */
+		data = buf;
+	}
+
 	/* take semaphore */
 	if (k_sem_take(&dev_data->write_lock, K_NO_WAIT)) {
 		return -EACCES;
```
Comment (tgorochowik) [inline drivers/flash/soc_flash_b91.c:155]: Should we free the buffer here?

File: drivers/flash/soc_flash_b91.c [hunk 5]
```diff
@@ -104,6 +135,21 @@ static int flash_b91_write(const struct device *dev, off_t offset,
 		return -EINVAL;
 	}
 
+	/* need to store data in intermediate RAM buffer in case from flash to flash write */
+	if ((uint32_t)data > FLASH_ORIGIN) {
```
Comment (tgorochowik) [inline drivers/flash/soc_flash_b91.c:139]: ``data >= FLASH_ORIGIN && data < (FLASH_ORIGIN + FLASH_SIZE)``

File: drivers/flash/soc_flash_b91.c (hunk 4 above)
Comment (yurvyn) [reply drivers/flash/soc_flash_b91.c]: Fixed.
Moved k_sem_take before k_malloc.

File: drivers/flash/soc_flash_b91.c (hunk 5 above)
Comment (yurvyn) [reply drivers/flash/soc_flash_b91.c]: Done.

**Review by nvlsianpu — DISMISSED** [review]

**Review by tgorochowik — APPROVED** [review]

File: drivers/flash/soc_flash_b91.c [hunk 6]
```diff
@@ -109,9 +140,30 @@ static int flash_b91_write(const struct device *dev, off_t offset,
 		return -EACCES;
 	}
 
+	/* need to store data in intermediate RAM buffer in case from flash to flash write */
+	if (((uint32_t)data >= FLASH_ORIGIN) &&
+		((uint32_t)data < (FLASH_ORIGIN + FLASH_SIZE))) {
+
+		buf = k_malloc(len);
+		if (buf == NULL) {
+			return -ENOMEM;
```
Comment (tgorochowik) [inline drivers/flash/soc_flash_b91.c:149]: Hmm, actually after this move we can exit here with the mutex locked, so it's still a problem

**Review by tgorochowik — CHANGES_REQUESTED** [review]

File: drivers/flash/soc_flash_b91.c (hunk 6 above)
Comment (yurvyn) [reply drivers/flash/soc_flash_b91.c:150]: Hmm, you are right.
Fixed.

**Review by nvlsianpu — APPROVED** [review]

**Review by tgorochowik — APPROVED** [review]

**Comment by yurvyn** [comment]
@carlescufi, @mbolivar-nordic, @cfriedt 
Please merge the PR.

**Comment by nvlsianpu** [comment]
@yurvyn This can be merged once zephyr v3.1.0 will be released.
